### PR TITLE
script/backport-create-issue: update tag custom field

### PR DESCRIPTION
**Pull request by batrick**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/script/backport-create-issue [hunk 1]
```diff
@@ -305,20 +305,37 @@ def mark_as_processed(r, issue):
...
 
+    logging.debug("custom_fields: %s", list(issue['custom_fields']))
+
+    try:
+        tags_cf = next(filter(lambda x: x['id'] == CF_TAGS, issue['custom_fields']))
+    except StopIteration:
+        tags_cf = {
+          'id': CF_TAGS,
+          'value': '',
+        }
```
Comment (epuertat) [inline src/script/backport-create-issue:317]: For a fallback assignment I find it easier to use `next(iter, default)`:

```suggestion
    tags_cf = next(
        filter(lambda x: x['id'] == CF_TAGS, issue['custom_fields']),
        default=dict(id=CF_TAGS, value='')
    )
```

**Review by epuertat — APPROVED** [review]
At first sight, I don't see anything wrong, but what's the context of this change? Can you plz add that to the PR and commit message? Thanks!

File: src/script/backport-create-issue (hunk 1 above)
Comment (batrick) [reply src/script/backport-create-issue]: Good suggestion, thanks.

**Comment by batrick** [comment]
@epuertat please take another look.

File: src/script/backport-create-issue [hunk 2]
```diff
@@ -305,20 +305,34 @@ def mark_as_processed(r, issue):
...
+            tags += f"{TAG_SEPARATOR}{TAG_BACKPORT_PROCESSED}"
+        else:
+            tags = TAG_BACKPORT_PROCESSED
+
+        logging.info("%s adding tag '%s', now '%s'", url(issue), TAG_BACKPORT_PROCESSED, tags)
+        tags_cf = {
+          'id': CF_TAGS,
+          'value': tags,
+        }
+        r.issue.update(issue.id, custom_fields=[tags_cf])
```
Comment (idryomov) [inline src/script/backport-create-issue:333]: Previously, the entire `custom_fields` list was passed here, not just the field that we are interested in updating.  Does Redmine preserve other custom fields on the ticket in this case?

File: src/script/backport-create-issue [hunk 3]
```diff
@@ -305,20 +305,34 @@ def mark_as_processed(r, issue):
...
-                             tag_backport_processed)
-                r.issue.update(issue.id, custom_fields=custom_fields)
-                return
 
+    logging.debug("custom_fields: %s", list(issue['custom_fields']))
+
+    tags_cf = next(filter(lambda x: x['id'] == CF_TAGS, issue['custom_fields']), None)
+
+    if tags_cf is None:
+        tags = ''
```
Comment (idryomov) [inline src/script/backport-create-issue:314]: Previously, tickets without the custom field that we are interested in were skipped.  Here, `tags = ''` and `if TAG_BACKPORT_PROCESSED not in tags` condition make the script attempt to set a custom field that isn't there.  Is that intended?

File: src/script/backport-create-issue (hunk 2 above)
Comment (batrick) [reply src/script/backport-create-issue:333]: Yes.

**Comment by batrick** [comment]
Merging this to get the script running again. If you have remaining concerns let's fix in a follow-up @idryomov

**Comment by epuertat** [comment]
@batrick I agree with @idryomov . Given custom field IDs seems to be randomly (re-)assigned on upgrades/migrations, I'd rather fetch it by name first, even if that involves an extra query on script run... Or at least set from env. var, so that it can be easily updated without further code changes.

**Comment by batrick** [comment]
> > If you have remaining concerns let's fix in a follow-up @idryomov
> 
> I'd have preferred if the script didn't hard-code the ID of the custom field and instead referred to it by name and fetched the ID from Redmine everytime the script is executed. But, as long as it works... ;)

The id should never change. I don't know why or how the id got changed by the tag plugin that was installed.

I prefer the "id" because any changes to the name (like what you are suggesting below) would break the script again.

> The other thing that I'd float around is renaming `Tags` field to `Tags (freeform)` to disambiguate it from the "real", now colorful, tags.

**Pull request merged**